src/main/kotlin/org/jitsi/nlj/RtpLayerDesc.kt

```diff
@@ -0,0 +1,294 @@
+import org.jitsi.nlj.stats.NodeStatsBlock
+import org.jitsi.utils.ArrayUtils
+import org.jitsi.utils.stats.RateStatistics
+import org.jitsi_modified.impl.neomedia.rtp.MediaSourceDesc
+
+/**
+ * Keeps track of its subjective quality index,
+ * its last stable bitrate and other useful things for adaptivity/routing.
+ *
+ * @author George Politis
+ */
+class RtpLayerDesc(
+    /**
+     * The [MediaSourceDesc] that this [RtpLayerDesc]
+     * belongs to.
+     */
+    private val source: MediaSourceDesc,
+    /**
+     * The index of this instance in the source layers array.
+     */
+    val index: Int,
+    /**
+     * The primary SSRC for this encoding.
+     */
+    val primarySSRC: Long,
+    /**
+     * The temporal layer ID of this instance.
+     */
+    val tid: Int,
+    /**
+     * The spatial layer ID of this instance.
+     */
+    val sid: Int,
+    /**
+     * The max height of the bitstream that this instance represents. The actual
+     * height may be less due to bad network or system load.
+     */
+    // XXX we should be able to sniff the actual height from the RTP
+    // packets.
+    val height: Int,
+    /**
+     * The max frame rate (in fps) of the bitstream that this instance
+     * represents. The actual frame rate may be less due to bad network or
+     * system load.
+     */
+    val frameRate: Double,
+    /**
+     * The [RtpLayerDesc] on which this layer depends.
+     */
+    private val dependencyLayers: Array<RtpLayerDesc>?
+) {
+    /**
+     * The ssrcs associated with this encoding (for example, RTX or FLEXFEC)
+     * Maps ssrc -> type [SsrcAssociationType] (rtx, etc.)
+     */
+    private val secondarySsrcs: MutableMap<Long, SsrcAssociationType> = HashMap()
+
+    /**
+     * The root [RtpLayerDesc] of the dependencies DAG. Useful for
+     * simulcast handling.
+     */
+    var baseLayer: RtpLayerDesc =
+        if (dependencyLayers == null || dependencyLayers.isEmpty())
+            this
+        else
+            dependencyLayers[0].baseLayer
+
+    /**
+     * The [RateStatistics] instance used to calculate the receiving
+     * bitrate of this RTP layer.
+     */
+    private val rateStatistics = RateStatistics(AVERAGE_BITRATE_WINDOW_MS)
+
+    /**
+     * Ctor.
+     *
+     * @param source the [MediaSourceDesc] that this instance
+     * belongs to.
+     * @param primarySSRC The primary SSRC for this layering/encoding.
+     */
+    constructor(
+        source: MediaSourceDesc,
+        primarySSRC: Long
+    ) : this(source, 0, primarySSRC, -1 /* tid */, -1 /* sid */,
+        NO_HEIGHT /* height */, NO_FRAME_RATE /* frame rate */,
+        null /* dependencies */)
+
+    /**
+     * @return the "id" of this layer/encoding. This is a server-side id and should
+     * not be confused with any encoding id defined in the client (such us the
+     * rid). This server-side id is used in the layer lookup table that is
+     * maintained in [MediaSourceDesc].
+     */
+    val encodingId: Long
```

Comment:
Is this scheme for the encoding ID new? ssrc + tid?

Reply:
No, this is just ported from the Java so far.  I'm not 100% sure what uses this ID and what uses the qualityId; one of my goals is to unify and/or clarify that.

Reply:
The encoding ID is exclusively used in the encodings cache for fast lookup. The quality is used in the adaptive track projection for filtering.